// react-resolver/dist/resolveError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

exports["default"] = resolve;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _Resolver = require("./Resolver");

var _Resolver2 = _interopRequireDefault(_Resolver);

function resolve(prop, catcher) {
  if (prop === undefined) prop = "error";

  return function resolveErrorDecorator(Component) {
    return (function (_React$Component) {
      _inherits(ResolveError, _React$Component);

      _createClass(ResolveError, null, [{
        key: "displayName",
        value: "ResolveError",
        enumerable: true
      }]);

      function ResolveError(props, context) {
        _classCallCheck(this, ResolveError);

        _get(Object.getPrototypeOf(ResolveError.prototype), "constructor", this).call(this, props, context);

        this.state = _defineProperty({}, prop, undefined);
      }

      _createClass(ResolveError, [{
        key: "handleCatch",
        value: function handleCatch(error) {
          this.setState(_defineProperty({}, prop, error));
        }
      }, {
        key: "handleResolve",
        value: function handleResolve(promise) {
          var _this = this;

          promise["catch"](function (error) {
            return _this.handleCatch(error);
          });
        }
      }, {
        key: "render",
        value: function render() {
          var _this2 = this;

          return _react2["default"].createElement(
            _Resolver2["default"],
            { onResolve: function (promise) {
                return _this2.handleResolve(promise);
              } },
            function (props) {
              return _react2["default"].createElement(Component, _extends({}, _this2.props, props, _this2.state));
            }
          );
        }
      }]);

      return ResolveError;
    })(_react2["default"].Component);
  };
}

module.exports = exports["default"];
